hack/update-codegen.sh

```diff
@@ -7,44 +7,55 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-GO_CMD=${1:-go}
 CURRENT_DIR=$(dirname "${BASH_SOURCE[0]}")
 TRAINING_OPERATOR_ROOT=$(realpath "${CURRENT_DIR}/..")
 TRAINING_OPERATOR_PKG="github.com/kubeflow/training-operator"
-CODEGEN_PKG=$(go list -m -mod=readonly -f "{{.Dir}}" k8s.io/code-generator)
 
 cd "$CURRENT_DIR/.."
 
-# shellcheck source=/dev/null
+# Get the code-generator binary.
+CODEGEN_PKG=$(go list -m -mod=readonly -f "{{.Dir}}" k8s.io/code-generator)
 source "${CODEGEN_PKG}/kube_codegen.sh"
+echo ">> Using ${CODEGEN_PKG}"
 
-# Generating conversion and defaults functions
+# Generating deepcopy and defaults.
+echo "Generating deepcopy and defaults for kubeflow.org/v1 and kubeflow.org/v2alpha1"
 kube::codegen::gen_helpers \
   --boilerplate "${TRAINING_OPERATOR_ROOT}/hack/boilerplate/boilerplate.go.txt" \
   "${TRAINING_OPERATOR_ROOT}/pkg/apis"
 
-# Generating OpenAPI for Kueue API extensions for v1
-kube::codegen::gen_openapi \
+# Generate clients for Training Operator V1 and V2
+echo "Generating clients for kubeflow.org/v1 and kubeflow.org/v2alpha1"
+kube::codegen::gen_client \
   --boilerplate "${TRAINING_OPERATOR_ROOT}/hack/boilerplate/boilerplate.go.txt" \
+  --output-dir "${TRAINING_OPERATOR_ROOT}/pkg/client" \
+  --output-pkg "${TRAINING_OPERATOR_PKG}/pkg/client" \
+  --with-watch \
+  --with-applyconfig \
+  "${TRAINING_OPERATOR_ROOT}/pkg/apis"
+
+# Get the kube-openapi binary.
+OPENAPI_PKG=$(go list -m -mod=readonly -f "{{.Dir}}" k8s.io/kube-openapi)
+echo ">> Using ${OPENAPI_PKG}"
+
+echo "Generating OpenAPI specification for kubeflow.org/v1"
+go run ${OPENAPI_PKG}/cmd/openapi-gen \
+  --go-header-file "${TRAINING_OPERATOR_ROOT}/hack/boilerplate/boilerplate.go.txt" \
   --output-pkg "${TRAINING_OPERATOR_PKG}/pkg/apis/kubeflow.org/v1" \
   --output-dir "${TRAINING_OPERATOR_ROOT}/pkg/apis/kubeflow.org/v1" \
+  --output-file "zz_generated.openapi.go" \
   --report-filename "${TRAINING_OPERATOR_ROOT}/hack/violation_exception_v1.list" \
-  --update-report \
   "${TRAINING_OPERATOR_ROOT}/pkg/apis/kubeflow.org/v1"
 
-# Generating OpenAPI for Kueue API extensions for v2alpha1
-kube::codegen::gen_openapi \
-  --boilerplate "${TRAINING_OPERATOR_ROOT}/hack/boilerplate/boilerplate.go.txt" \
+echo "Generating OpenAPI specification for kubeflow.org/v2alpha1"
+go run ${OPENAPI_PKG}/cmd/openapi-gen \
+  --go-header-file "${TRAINING_OPERATOR_ROOT}/hack/boilerplate/boilerplate.go.txt" \
   --output-pkg "${TRAINING_OPERATOR_PKG}/pkg/apis/kubeflow.org/v2alpha1" \
   --output-dir "${TRAINING_OPERATOR_ROOT}/pkg/apis/kubeflow.org/v2alpha1" \
+  --output-file "zz_generated.openapi.go" \
   --report-filename "${TRAINING_OPERATOR_ROOT}/hack/violation_exception_v2alpha1.list" \
-  --update-report \
   "${TRAINING_OPERATOR_ROOT}/pkg/apis/kubeflow.org/v2alpha1"
```

Comment:
Actually, I was thinking a bit more on this and I suggest that we rely on `kube-openapi` to generate the OpenAPI spec for our APIs.
I don't want our OpenAPI definitions to contain non-Kubeflow APIs like `k8s.io/apimachinery`, since Kubernetes has their own OpenAPI spec for it.

Which means users should:
- `import kubeflow.training.models` for Kubeflow APIs
- `import jobset.models` for JobSet APIs
- `import kubernetes.client.models` for Kubernetes APIs.

Additionally, in our SDK docs and models users can see only Kubeflow-related APIs: https://github.com/kubeflow/training-operator/tree/9bea4cfaeed9c17833760561e2605bba7509bba2/sdk_v2/docs

I can see that JobSet also uses the `kube-openapi` for it: https://github.com/kubernetes-sigs/jobset/blob/main/Makefile#L244-L245.


WDYT @tenzen-y @kubeflow/wg-training-leads @astefanutti ?

Reply:
Yes I agree. I myself wasn't satisfied either with the `kube_codegen.sh` script pulling these `k8s.io/apimachinery/pkg/{apis,runtime,version}` packages, and it feels wrong from the end-user standpoint to have these redefined in the Kubeflow APIs.